### ISSUE TITLE
Add better meta-layer approach

### ISSUE THREAD

**Issue by allisonkarlitskaya** [containers/composefs-rs#21]
Instead of returning the second-highest layer as the meta-layer, we can:
 - explicitly label the meta layer as a label on the container config
 - explicitly label the UKI as a label on the container config (probably by its fs-verity digest)
 - have some kind of a nicer `install-kernel` command or so that defaults to unpacking in `/boot`.  We could use that during the OS image build, but it would also allow for easier "upgrades" on running systems.

**Comment by allisonkarlitskaya** [comment]
The issue with just putting it in the filesystem is the circular hash dependency thing.  It needs to be out of band, either in an artifact or hidden layer.  That's the thing that needs to be identified.

In the hidden layer, I was actually imagining something like following the bootloader spec.

I started to lean against the idea of hardcoding UKI and having a verity digest for that one file.  Following BLS actually gives us the ability to also do this without a UKI, which seems like something that someone might like to do.

**Comment by cgwalters** [comment]
In what I'm thinking here, we do:

- build rootfs (without `/usr/lib/modules/$kver/uki.efi`), get its fsverity digest, embed in uki
- Add a new layer with that UKI
- compute the *standard* fsverity digest (i.e. the non-bootable generic container one), add as label to the config as usual

When we boot it'd be into the rootfs without the uki, but it'd be in that regular place when we're run as a container image.

What I don't think we've talked about though is whether and how we want to try to create a chain that allows us to establish trust in the manifest/config starting from the rootfs. I think in the longer term we do need to do that in order to have things like `bootc status` as well as general `bootc upgrade` doing change detection (i.e. no-op upgrades if we're on a matching manifest digest already). We could start by saying that's just local unsigned state, but I think we can do it with something like https://github.com/containers/composefs/issues/294#issuecomment-2299142448

**Comment by cgwalters** [comment]
> That means that it's going to be the "wrong" fs-verity compared to the one that we actually want to mount.

Yes for sure, we'd want a special annotation on the final kernel layer, so that tooling knows to extract and also make a composefs for everything before that layer as well.

I don't see this as weird at all - remember in a general image derivation case we'll often have around a composefs for layers 0..N for a base image and another for layers 0..(N+1) for a final derived image.

> The "hidden metadata layer" approach is a massively gigantic hack

Can you write out a bit more what this is? There's not many comments in the code and it's not immediately obvious to me how it works in the code right now.

**Comment by cgwalters** [comment]
OK, right I think I remember you describing this "add then whiteout". I need to think harder about the security properties of this versus my proposal of "just add a layer".

First, do you agree that "bootable verified OCI" should just be a special case of "verified OCI" for apps or other use cases? If so, then in the "add then whiteout" we would have a situation where the config digest == UKI digest right? 
But when we have the manifest layer annotations as I think we should, then we automatically will have a verifiable digest covering the "meta-layer" as you call it (the non-whiteout one) and given the config we can then find a verity digest of the UKI to install it.

But in the end what's the advantage of having "config digest == UKI digest"? We can equally well access the UKI in a "just add a layer" approach - here config digest != UKI digest, but I don't see a problem with that; from the system again given a verified config, finding the UKI is just traversing into its standard place in `/usr/lib/modules` instead of finding the meta-layer.

So they seem equivalent from a security PoV. From an elegance/understandable PoV, wouldn't you agree it's just nicer to have e.g. `podman|docker run <image>` let you see the UKI?

**Comment by allisonkarlitskaya** [comment]
I think I want to do #35.  It's the least magical way.
